tests/test_slurm_system.py

```diff
@@ -142,3 +143,88 @@ def test_update_node_states_with_mocked_outputs(mock_get_sinfo, mock_get_squeue,
     parsed_node_list = parse_node_list(node_list)
     assert parsed_node_list == expected_parsed_node_list
+
+
+def setup_mock_slurm_system(slurm_system, group_name):
+    """
+    Helper function to set up a mock Slurm system with nodes and their states.
+    Args:
+        slurm_system: The Slurm system instance.
+        group_name: Name of the group to set up within the partition.
+    """
+    partition_name = "backup"
+
+    mock_nodes = [
+        SlurmNode(name="node01", partition=partition_name, state=SlurmNodeState.IDLE),
+        SlurmNode(name="node02", partition=partition_name, state=SlurmNodeState.COMPLETING),
+        SlurmNode(name="node03", partition=partition_name, state=SlurmNodeState.IDLE),
+        SlurmNode(name="node04", partition=partition_name, state=SlurmNodeState.DOWN),
+    ]
+
+    slurm_system.partitions = [
+        SlurmPartition(
+            name=partition_name,
+            nodes=[node.name for node in mock_nodes],
+            groups=[SlurmGroup(name=group_name, nodes=[node.name for node in mock_nodes])],
+        )
+    ]
+
+    slurm_system.partitions[0]._slurm_nodes = mock_nodes
+
+    grouped_nodes = {
+        SlurmNodeState.IDLE: [
+            slurm_system.groups[partition_name][group_name][0],
+            slurm_system.groups[partition_name][group_name][2],
+        ],
+        SlurmNodeState.COMPLETING: [
+            slurm_system.groups[partition_name][group_name][1],
+        ],
+        SlurmNodeState.ALLOCATED: [],
+    }
+
+    return grouped_nodes
+
+
+def test_allocate_nodes_max_avail(slurm_system):
+    group_name = "backup"
+
+    grouped_nodes = setup_mock_slurm_system(slurm_system, group_name)
+
+    available_nodes = slurm_system.allocate_nodes(grouped_nodes, "max_avail", group_name)
+    expected_node_names = [
+        slurm_system.groups["backup"][group_name][0].name,
+        slurm_system.groups["backup"][group_name][1].name,
+        slurm_system.groups["backup"][group_name][2].name,
+    ]
+    returned_node_names = [node.name for node in available_nodes]
+
+    assert set(returned_node_names) == set(expected_node_names), "Should return all available nodes except DOWN nodes"
+    assert slurm_system.groups["backup"][group_name][3] not in returned_node_names, "DOWN node should not be included"
+
+
+def test_allocate_nodes_num_nodes_integers(slurm_system):
+    group_name = "backup"
+
+    grouped_nodes = setup_mock_slurm_system(slurm_system, group_name)
+
+    max_available_nodes = slurm_system.allocate_nodes(grouped_nodes, 2, group_name)
+    expected_node_names = [
+        slurm_system.groups["backup"][group_name][0].name,
+        slurm_system.groups["backup"][group_name][2].name,
+    ]
+    returned_node_names = [node.name for node in max_available_nodes]
+
+    assert set(returned_node_names) == set(expected_node_names), "Should return 2 available nodes"
+    assert slurm_system.groups["backup"][group_name][1] not in returned_node_names, "Only 2 nodes should be included"
```

Comment:
Need to add `.name`, otherwise this assert is checking `object not in list[str]`.